components/external_storage/src/s3.rs

```diff
@@ -0,0 +1,208 @@
+        // than 100MB. See https://github.com/rusoto/rusoto/pull/1227
+        // for more information.
+        let mut http_config = HttpConfig::new();
+        http_config.read_buf_size(1024 * 1024 * 2);
+        let http_dispatcher = HttpClient::new_with_config(http_config).unwrap();
+
+        S3Storage::with_request_dispatcher(config, http_dispatcher)
+    }
+
+    fn with_request_dispatcher<D>(config: &Config, dispatcher: D) -> Result<S3Storage>
+    where
+        D: DispatchSignedRequest + Send + Sync + 'static,
+        D::Future: Send,
+    {
+        if config.bucket.is_empty() {
+            return Err(Error::new(ErrorKind::InvalidInput, "missing bucket name"));
+        }
+        let region = if config.endpoint.is_empty() {
+            config.region.parse::<region::Region>().map_err(|e| {
+                Error::new(
+                    ErrorKind::InvalidInput,
+                    format!("invalid region format {}: {}", config.region, e),
+                )
+            })?
+        } else {
+            region::Region::Custom {
+                name: config.region.clone(),
+                endpoint: config.endpoint.clone(),
+            }
+        };
+        let client = if config.access_key.is_empty() || config.secret_access_key.is_empty() {
+            let cred_provider = DefaultCredentialsProvider::new().map_err(|e| {
+                Error::new(
+                    ErrorKind::PermissionDenied,
+                    format!("unable to get credentials: {}", e),
+                )
+            })?;
+            S3Client::new_with(dispatcher, cred_provider, region)
+        } else {
+            let cred_provider = StaticProvider::new(
+                config.access_key.clone(),
+                config.secret_access_key.clone(),
+                None, /* token */
+                None, /* valid_for */
+            );
+            S3Client::new_with(dispatcher, cred_provider, region)
+        };
+        Ok(S3Storage {
+            config: config.clone(),
+            client,
+        })
+    }
+
+    fn maybe_prefix_key(&self, key: &str) -> String {
+        if !self.config.prefix.is_empty() {
+            return format!("{}/{}", self.config.prefix, key);
+        }
+        key.to_owned()
+    }
+}
+
+impl ExternalStorage for S3Storage {
+    fn write(
+        &self,
+        name: &str,
+        reader: &mut (dyn AsyncRead + Unpin + Send + 'static),
+    ) -> Result<()> {
+        let key = self.maybe_prefix_key(name);
+        debug!("save file to s3 storage"; "key" => %key);
+        let get_var = |s: &String| {
+            if s.is_empty() {
+                None
+            } else {
+                Some(s.clone())
+            }
+        };
+        let req = PutObjectRequest {
+            key,
+            bucket: self.config.bucket.clone(),
+            body: Some(ByteStream::new(
+                FramedRead::new(reader.compat(), BytesCodec::new()).map(|bytes| Bytes::from(bytes)),
```

Comment:
@kennytm I updated the code here to convert `AsyncRead` to futures 0.1 `Stream`, but still getting compile error. `ByteStream::new` require the input `Stream` to have `Send + 'static`. I added `Send + 'static` as trait bound to `reader`, but gets the following compile error about conflicting lifetime. Am I doing anything wrong, and any idea to get around the error?
<img width="1619" alt="Screen Shot 2020-01-12 at 5 23 05 PM" src="https://user-images.githubusercontent.com/2606959/72216746-3cb74e00-3560-11ea-8cd1-5dadd829e5d6.png">

Reply:
The lifetime error is about `reader`, not `*reader`. In principle you need `&'static mut T` for this to work.

Could we change the type of `reader` to `Box<dyn AsycnRead + Unpin + Send>`?

BTW you could use `bytes.freeze()` instead of `Bytes::from(bytes)`.

Reply:
Thanks much. It works. I'll update the code soon.